arangod/Aql/ExecutionEngine.cpp

```diff
@@ -257,16 +259,26 @@ ExecutionEngine::~ExecutionEngine() {
   if (_sharedState) {  // ensure no async task is working anymore
     _sharedState->invalidate();
   }
+
+  if (_asyncPrefetchSlotsReserved > 0) {
+    // return leased slots to QueryRegistryFeature
+    _query.vocbase()
+        .server()
+        .getFeature<QueryRegistryFeature>()
+        .returnAsyncPrefetchSlots(_asyncPrefetchSlotsReserved);
```

Comment:
Can we store a reference to the feature instead of using `getFeature` here? Ideally we would take a reference to the feature in the constructor, but I would also accept a single `getFeature` call in the constructor to populate the member.

Actually, after think about this a bit more, I think this tracking logic should not be part of the `QueryRegistryFeature` at all. I would rather put it into a separate class, and have the ExecutionEngine keep a reference to that class. The `QueryRegistryFeature` can still be the owner of that instance (since it also defines the program options, and we currently have no simple way to move them somehwere that is not a feature), but that way the ExecutionEngine at least has no dependency on the the QueryRegistryFeature.

Even better would be to use RAII and have a guard class that returns the slots,